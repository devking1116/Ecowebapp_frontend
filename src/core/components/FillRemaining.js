/* @flow */
import React from 'react';
import reactCSS from 'reactcss';

type Props = {
  direction?: 'column' | 'row',
  children?: any,
};

/**
 * A component that will keep it's first child it's given size, but grow and shrink it's remaining child to
 * fill the entier component.
 */
export default function FillRemaining({ direction, children, ...otherProps }: Props): React.Element<*> {
  if(React.Children.count(children) < 2) throw new Error('FillRemaining must contain more then one child.');
  const [ firstChild, ...remainingChildren ] = React.Children.toArray(children);

  if(!direction || direction === 'column') {
    return (
      <FillRemainingColumn fixedChild={firstChild} {...otherProps}>
        {remainingChildren}
      </FillRemainingColumn>
    );
  }
  else if(direction === 'row') {
    return (
      <FillRemainingRow fixedChild={firstChild} {...otherProps}>
        {remainingChildren}
      </FillRemainingRow>
    );
  }
  else {
    throw new Error(`Invalid direction, ${direction}, for FillRemaining.`);
  }
}

function FillRemainingColumn({ fixedChild, children, ...otherProps }: *): React.Element<*> {
  const styles = reactCSS({
    'default': {
      wrapper: {
        display: 'flex',
        flexDirection: 'column',
        width: '100%',
        height: '100%',
      },
      fixed: {
      },
      stretch: {
        display: 'flex',
        flexDirection: 'row',
        flex: '1 0 1px',
      },
      fillWrapper: {
        position: 'relative',
        width: '100%',
      },
      fill: {
        position: 'absolute',
        top: 0,
        left: 0,
        bottom: 0,
        right: 0,
      },
    },
  });

  return (
    <div style={styles.wrapper} {...otherProps}>
      <div style={styles.fixed}>
        {fixedChild}
      </div>
      <div style={styles.stretch}>
        <div style={styles.fillWrapper}>
          <div style={styles.fill}>
            {children}
          </div>
        </div>
      </div>
    </div>
  );
}

function FillRemainingRow({ fixedChild, children, ...otherProps }: *): React.Element<*> {
  const styles = reactCSS({
    'default': {
      wrapper: {
        display: 'flex',
        flexDirection: 'row',
        width: '100%',
        height: '100%',
      },
      fixed: {
        height: '100%',
      },
      stretch: {
        flex: '1 0 1px',
        height: '100%',
      },
    },
  });

  return (
    <div style={styles.wrapper} {...otherProps}>
      <div style={styles.fixed} >
        {fixedChild}
      </div>
      <div style={styles.stretch}>
        {children}
      </div>
    </div>
  );
}
