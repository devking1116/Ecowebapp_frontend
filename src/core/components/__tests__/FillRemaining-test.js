import React from 'react';
import renderer from 'react-test-renderer';

import FillRemaining from '../FillRemaining';

test('it uses the flex-direction given by direction prop.', () => {
  const componentRow = renderer.create(
    <FillRemaining direction="row">
      <div>First Child</div>
      <div>Second Child</div>
      <div>Third Child</div>
    </FillRemaining>
  );
  const componentColumn = renderer.create(
    <FillRemaining direction="column">
      <div>First Child</div>
      <div>Second Child</div>
      <div>Third Child</div>
    </FillRemaining>
  );

  const treeRow = componentRow.toJSON();
  const treeColumn = componentColumn.toJSON();

  expect(treeRow).toMatchSnapshot();
  expect(treeColumn).toMatchSnapshot();
});

test('it uses column as the flex-direction if the direction prop is not given.', () => {
  const component = renderer.create(
    <FillRemaining>
      <div>First Child</div>
      <div>Second Child</div>
      <div>Third Child</div>
    </FillRemaining>
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('it throws an error when called with one child.', () => {
  const testRender = () => {
    renderer.create(
      <FillRemaining>
        <div>Only Child</div>
      </FillRemaining>
    );
  };

  expect(testRender).toThrowErrorMatchingSnapshot();
});

test('it throws an error when called with one child.', () => {
  const testRender = () => {
    renderer.create(
      <FillRemaining direction="not-row-or-column">
        <div>First Child</div>
        <div>Second Child</div>
      </FillRemaining>
    );
  };

  expect(testRender).toThrowErrorMatchingSnapshot();
});
