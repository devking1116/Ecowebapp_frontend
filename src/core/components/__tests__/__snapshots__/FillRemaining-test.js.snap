exports[`test it throws an error when called with one child. 1`] = `"FillRemaining must contain more then one child."`;

exports[`test it throws an error when called with one child. 2`] = `"Invalid direction, not-row-or-column, for FillRemaining."`;

exports[`test it uses column as the flex-direction if the direction prop is not given. 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "width": "100%",
    }
  }>
  <div
    style={Object {}}>
    <div>
      First Child
    </div>
  </div>
  <div
    style={
      Object {
        "MozBoxFlex": "1 0 1px",
        "WebkitBoxFlex": "1 0 1px",
        "WebkitFlex": "1 0 1px",
        "display": "flex",
        "flex": "1 0 1px",
        "flexDirection": "row",
        "msFlex": "1 0 1px",
      }
    }>
    <div
      style={
        Object {
          "position": "relative",
          "width": "100%",
        }
      }>
      <div
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }>
        <div>
          Second Child
        </div>
        <div>
          Third Child
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test it uses the flex-direction given by direction prop. 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "height": "100%",
      "width": "100%",
    }
  }>
  <div
    style={
      Object {
        "height": "100%",
      }
    }>
    <div>
      First Child
    </div>
  </div>
  <div
    style={
      Object {
        "MozBoxFlex": "1 0 1px",
        "WebkitBoxFlex": "1 0 1px",
        "WebkitFlex": "1 0 1px",
        "flex": "1 0 1px",
        "height": "100%",
        "msFlex": "1 0 1px",
      }
    }>
    <div>
      Second Child
    </div>
    <div>
      Third Child
    </div>
  </div>
</div>
`;

exports[`test it uses the flex-direction given by direction prop. 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "width": "100%",
    }
  }>
  <div
    style={Object {}}>
    <div>
      First Child
    </div>
  </div>
  <div
    style={
      Object {
        "MozBoxFlex": "1 0 1px",
        "WebkitBoxFlex": "1 0 1px",
        "WebkitFlex": "1 0 1px",
        "display": "flex",
        "flex": "1 0 1px",
        "flexDirection": "row",
        "msFlex": "1 0 1px",
      }
    }>
    <div
      style={
        Object {
          "position": "relative",
          "width": "100%",
        }
      }>
      <div
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }>
        <div>
          Second Child
        </div>
        <div>
          Third Child
        </div>
      </div>
    </div>
  </div>
</div>
`;
